File: code/ex148.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {    // O(nlog(n))
        if (head == null || head.next == null)    return head;
        
        ListNode[] listNodeArray = divide(head);
        ListNode leftHead = listNodeArray[0], rightHead = listNodeArray[1];
        
        // Divide
        ListNode left = sortList(leftHead);
        ListNode right = sortList(rightHead);
               
        // Conquer
        return merge(left, right);
    }
    
    private ListNode[] divide(ListNode head) {    // O(n)
        // Get length
        ListNode p = head;
        int len = 0;
        while (p != null) {
            len++;
            p = p.next;
        }
        
        // Get the tail position of the left half linked list
        int l1 = len / 2;
        p = head;
        while (l1 > 1) {
            p = p.next;
            l1--;
        }
        
        // Break linked list
        ListNode h1 = head;
        ListNode h2 = p.next;
        p.next = null;
        
        return new ListNode[]{h1, h2};
    }
    
    private ListNode merge(ListNode left, ListNode right) {    // O(n)
        ListNode head;
        if (left.val < right.val) {
            head = left;
            left = left.next;
        } else {
            head = right;
            right = right.next;
        }
        
        ListNode p = head;
        while (left != null || right != null) {
            if (left == null) {
                p.next = right;
                p = p.next;
                right = right.next;
            }
            else if (right == null) {
                p.next = left;
                p = p.next;
                left = left.next;
            }
            else {
                if (left.val < right.val) {
                    p.next = left;
                    p = p.next;
                    left = left.next;
                } else {
                    p.next = right;
                    p = p.next;
                    right = right.next;
                }
            }
        }
        p.next = null;
            
        return head;
    }
}

File: code/ex148.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        if (head == NULL || head->next == NULL)    return head;
        
        // Find the head of two half of the linkedlist
        ListNode* first = head;    ListNode* second = head;
        int len = 0;
        for (ListNode* p = head; p != NULL; p = p->next)    len++;
        for (int i = 0; i < len / 2 - 1; i++)    second = second->next;
        ListNode* dummy = second;    second = second->next;    dummy->next = NULL;
        
        // Divide
        ListNode* sorted1 = sortList(first);
        ListNode* sorted2 = sortList(second);
        
        // Conquer
        ListNode* sorted = new ListNode(0);
        ListNode* p = sorted;    ListNode* p1 = sorted1;    ListNode* p2 = sorted2;
        while (p1 != NULL || p2 != NULL) {
            if (p1 == NULL) {
                p->next = p2;
                p2 = p2->next;
            } else if (p2 == NULL) {
                p->next = p1;
                p1 = p1->next;
            } else if (p1->val < p2->val) {
                p->next = p1;
                p1 = p1->next;
            } else {
                p->next = p2;
                p2 = p2->next;
            }
            p = p->next;
        }
        
        return sorted->next;
    }
};